README.md
# go-money A monetary type for go

This repo is to be a generic money type for golang.  It associates a currency with the amount, and the default implementation uses int64 as the underlying datatype.  Input and output is via a string with a "." as the decimal point.

![Go](https://github.com/314159/go-money/workflows/Go/badge.svg)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changes
- Upgrade github.com/stretchr/testify to v1.8.0
- Upgrade go version used by github actions to v1.18.x
- Allow trailing decimal zeros.  Fixes issue [#4](https://github.com/314159/go-money/issues/4)
- No longer creating "preemptive interfaces"

## [v1.0.2]
### SECURITY
- Upgrade yaml.v3 module

## [v1.0.1]
### Changes
- Move decimal point to a constant in prep to make it configurable in the future.

## [v1.0.0]
### Added
- Initial release

<!-- markdownlint-configure-file { "MD022": false, "MD024": false, "MD030": false, "MD032": false} -->


money/money_test.go
package money_test

import (
	"testing"

	"github.com/314159/go-money/currency"
	"github.com/314159/go-money/money"

	"github.com/stretchr/testify/assert"
)

func TestMoney(t *testing.T) {
	usDollar := currency.New("USD", "840", 2)
	jpYen := currency.New("JPY", "139", 0)

	testCases := []struct {
		Case     string
		Amount   string
		Currency money.Currency
		Expects  string
		ExpCurr  money.Currency
		ExpErr   bool
		ErrMsg   string
	}{
		{
			Case:     "123 USD",
			Amount:   "123",
			Currency: usDollar,
			Expects:  "123.00",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "-123 USD",
			Amount:   "-123",
			Currency: usDollar,
			Expects:  "-123.00",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "-.01 USD",
			Amount:   "-.01",
			Currency: usDollar,
			Expects:  "-0.01",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "-.11 USD",
			Amount:   "-.11",
			Currency: usDollar,
			Expects:  "-0.11",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "-0.11 USD",
			Amount:   "-0.11",
			Currency: usDollar,
			Expects:  "-0.11",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "empty",
			Amount:   "",
			Currency: usDollar,
			Expects:  "0.00",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "123.10 USD",
			Amount:   "123.10",
			Currency: usDollar,
			Expects:  "123.10",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "123.1 USD",
			Amount:   "123.1",
			Currency: usDollar,
			Expects:  "123.10",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "123.123 USD",
			Amount:   "123.123",
			Currency: usDollar,
			Expects:  "",
			ExpCurr:  nil,
			ExpErr:   true,
			ErrMsg:   "invalid Money amount: 123.123",
		},
		{
			Case:     "123.12.3 USD",
			Amount:   "123.12.3",
			Currency: usDollar,
			Expects:  "",
			ExpCurr:  nil,
			ExpErr:   true,
			ErrMsg:   "invalid Money amount: 123.12.3",
		},
		{
			Case:     "Non-numeric-amount",
			Amount:   "12.3X",
			Currency: usDollar,
			Expects:  "",
			ExpCurr:  nil,
			ExpErr:   true,
			ErrMsg:   "invalid Money amount: 12.3X",
		},
		{
			Case:     "123 JPY",
			Amount:   "123",
			Currency: jpYen,
			Expects:  "123",
			ExpCurr:  jpYen,
			ExpErr:   false,
		},
		{
			Case:     "Extra zeros in decimal part",
			Amount:   "123.4500",
			Currency: usDollar,
			Expects:  "123.45",
			ExpCurr:  usDollar,
			ExpErr:   false,
		},
		{
			Case:     "123.10 JPY",
			Amount:   "123.10",
			Currency: jpYen,
			Expects:  "",
			ExpCurr:  nil,
			ExpErr:   true,
			ErrMsg:   "invalid Money amount: 123.10",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.Case, func(t *testing.T) {
			var actual string
			var c money.Currency

			m, err := money.NewMoney(tc.Amount, tc.Currency)
			if m != nil {
				actual = m.String()
				c = m.Currency()
			}

			hadError := err != nil
			var es string

			if err != nil {
				es = err.Error()
			}

			assert.Equal(t, tc.Expects, actual)
			assert.Equal(t, tc.ExpCurr, c)
			assert.Equal(t, tc.ExpErr, hadError)
			assert.Equal(t, tc.ErrMsg, es)
		})
	}
}


money/currency_test.go
package money_test

import (
	"testing"

	"github.com/314159/go-money/money"
	"github.com/stretchr/testify/assert"
)

func TestAddCurrency(t *testing.T) {
	testCases := []struct {
		testName  string
		code      string
		c         money.Currency
		exists    bool
		expectNil bool
	}{
		{
			testName:  "New Currency",
			code:      "AAA",
			c:         money.NewCurrency("AAA", "000", 4),
			exists:    false,
			expectNil: false,
		},
		{
			testName:  "Try Add nil currency",
			code:      "NIL",
			c:         nil,
			exists:    false,
			expectNil: true,
		},
		{
			testName:  "Replace Currency",
			code:      "AAA",
			c:         money.NewCurrency("AAA", "111", 4),
			exists:    true,
			expectNil: false,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.testName, func(t *testing.T) {
			before := money.FindCurrency(tc.code)
			money.AddCurrency(tc.c)
			after := money.FindCurrency(tc.code)

			assert.Equal(t, tc.exists, before != nil)
			assert.Equal(t, tc.expectNil, after == nil)
		})
	}
}
func TestFindCurrency(t *testing.T) {
	testCases := []struct {
		Case  string
		code  string
		s     string
		isNil bool
		nc    string
		dd    int
	}{
		{
			Case:  "US Dollars",
			code:  "USD",
			isNil: false,
			s:     "USD",
			nc:    "840",
			dd:    2,
		},
		{
			Case:  "Euro Dollars",
			code:  "EUR",
			isNil: false,
			s:     "EUR",
			nc:    "978",
			dd:    2,
		},
		{
			Case:  "Great Brittan Pounds",
			code:  "GBP",
			isNil: false,
			s:     "GBP",
			nc:    "826",
			dd:    2,
		},
		{
			Case:  "Japanese Yen",
			code:  "JPY",
			isNil: false,
			s:     "JPY",
			nc:    "392",
			dd:    0,
		},
		{
			Case:  "Romulan Ale",
			code:  "AAZ",
			isNil: true,
			s:     "",
			nc:    "",
			dd:    0,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.Case, func(t *testing.T) {
			var s string
			var nc string
			var dd int

			c := money.FindCurrency(tc.s)
			if c != nil {
				s = c.String()
				nc = c.NumericCode()
				dd = c.DecimalDigits()
			}

			assert.Equal(t, tc.isNil, c == nil)
			assert.Equal(t, tc.s, s)
			assert.Equal(t, tc.nc, nc)
			assert.Equal(t, tc.dd, dd)
		})
	}
}


money/money.go
package money

import (
	"fmt"
	"strconv"
	"strings"
)

// Currency represents a monetary currency
type Currency interface {
	NumericCode() string
	DecimalDigits() int
	String() string
}

type money struct {
	value    int64
	currency Currency
}

const decimalPoint = "."

func (m money) String() string {
	dp := m.currency.DecimalDigits()
	av, sign := func(v int64) (int64, string) {
		sign := ""
		if v < 0 {
			sign = "-"
			v = v * -1
		}
		return v, sign
	}(m.value)
	v := strconv.FormatInt(av, 10)

	switch {
	case dp == 0:
		return fmt.Sprintf("%s%s", sign, v)
	case len(v) < dp:
		return fmt.Sprintf("%s0.%0*s", sign, dp-len(v)+1, v)
	case len(v) == dp:
		return fmt.Sprintf("%s0.%s", sign, v)
	default:
		integerPart := v[:len(v)-dp]
		decimalPart := v[len(v)-dp:]
		return fmt.Sprintf("%s%s%s%s", sign, integerPart, decimalPoint, decimalPart)
	}
}
func (m money) Currency() Currency {
	return m.currency
}

// NewMoney constructs a Money object from the amount string and the currency.
func NewMoney(amount string, c Currency) (*money, error) {
	parts := strings.Split(amount, decimalPoint)

	var integerPart, decimalPart string

	switch {
	case len(parts) == 1:
		integerPart = parts[0]
		decimalPart = ""
	case len(parts) == 2:
		integerPart = parts[0]
		decimalPart = parts[1]
	default:
		return nil, fmt.Errorf("invalid Money amount: %s", amount)
	}

	l := len(decimalPart)
	dd := c.DecimalDigits()

	switch {
	case l < dd:
		decimalPart = decimalPart + fmt.Sprintf("%0*d", dd-l, 0)
	case l == dd:
		break
	case (l > dd) && (decimalPart[dd:] == strings.Repeat("0", l-dd)):
		decimalPart = decimalPart[:dd]
	default:
		// This would cause loss of precision!
		return nil, fmt.Errorf("invalid Money amount: %s", amount)
	}

	v, err := strconv.ParseInt(integerPart+decimalPart, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid Money amount: %s", amount)
	}

	m := money{value: v, currency: c}
	return &m, nil
}


currencylist/currencylist.go
package currencylist

import (
	"github.com/314159/go-money/currency"
	"github.com/314159/go-money/money"
)

// currencyList is a map of currencies
type currencyList struct {
	list map[string]money.Currency
}

func New() *currencyList {
	return &currencyList{
		list: make(map[string]money.Currency, 0),
	}
}

// AddCurrency will add a currency to the list of currencies you can search for.
func (l *currencyList) AddCurrency(c money.Currency) {
	if c == nil {
		return
	}

	l.list[c.String()] = c
	l.list[c.NumericCode()] = c
}

// FindCurrency will return a currency based on the currency code (numeric or alpha)
func (l *currencyList) FindCurrency(code string) money.Currency {
	c, ok := l.list[code]

	if !ok {
		return nil
	}
	return c
}

// AddMany will add a slice of Currencys to the list
func (l *currencyList) AddMany([]money.Currency) {
	for _, c := range currency.StandardList {
		l.AddCurrency(c)
	}
}


currency/currency.go
package currency

import "github.com/314159/go-money/money"

type currency struct {
	alphaCode     string
	numericCode   string
	decimalDigits int
}

func (c currency) AlphaCode() string {
	return c.alphaCode
}

func (c currency) NumericCode() string {
	return c.numericCode
}

func (c currency) DecimalDigits() int {
	return c.decimalDigits
}

func (c currency) String() string {
	return c.AlphaCode()
}

// New constructs a currency object
func New(alphaCode, numericCode string, decimalDigits int) *currency {
	c := currency{
		alphaCode:     alphaCode,
		numericCode:   numericCode,
		decimalDigits: decimalDigits,
	}
	return &c
}

var StandardList = []money.Currency{
	// list retrieved 2020-11-23 from https://en.wikipedia.org/wiki/ISO_4217
	New("AED", "784", 2),
	New("AFN", "971", 2),
	New("ALL", "008", 2),
	New("AMD", "051", 2),
	New("ANG", "532", 2),
	New("AOA", "973", 2),
	New("ARS", "032", 2),
	New("AUD", "036", 2),
	New("AWG", "533", 2),
	New("AZN", "944", 2),
	New("BAM", "977", 2),
	New("BBD", "052", 2),
	New("BDT", "050", 2),
	New("BGN", "975", 2),
	New("BHD", "048", 3),
	New("BIF", "108", 0),
	New("BMD", "060", 2),
	New("BND", "096", 2),
	New("BOB", "068", 2),
	New("BOV", "984", 2),
	New("BRL", "986", 2),
	New("BSD", "044", 2),
	New("BTN", "064", 2),
	New("BWP", "072", 2),
	New("BYN", "933", 2),
	New("BZD", "084", 2),
	New("CAD", "124", 2),
	New("CDF", "976", 2),
	New("CHE", "947", 2),
	New("CHF", "756", 2),
	New("CHW", "948", 2),
	New("CLF", "990", 4),
	New("CLP", "152", 0),
	New("CNY", "156", 2),
	New("COP", "170", 2),
	New("COU", "970", 2),
	New("CRC", "188", 2),
	New("CUC", "931", 2),
	New("CUP", "192", 2),
	New("CVE", "132", 2),
	New("CZK", "203", 2),
	New("DJF", "262", 0),
	New("DKK", "208", 2),
	New("DOP", "214", 2),
	New("DZD", "012", 2),
	New("EGP", "818", 2),
	New("ERN", "232", 2),
	New("ETB", "230", 2),
	New("EUR", "978", 2),
	New("FJD", "242", 2),
	New("FKP", "238", 2),
	New("GBP", "826", 2),
	New("GEL", "981", 2),
	New("GHS", "936", 2),
	New("GIP", "292", 2),
	New("GMD", "270", 2),
	New("GNF", "324", 0),
	New("GTQ", "320", 2),
	New("GYD", "328", 2),
	New("HKD", "344", 2),
	New("HNL", "340", 2),
	New("HRK", "191", 2),
	New("HTG", "332", 2),
	New("HUF", "348", 2),
	New("IDR", "360", 2),
	New("ILS", "376", 2),
	New("INR", "356", 2),
	New("IQD", "368", 3),
	New("IRR", "364", 2),
	New("ISK", "352", 0),
	New("JMD", "388", 2),
	New("JOD", "400", 3),
	New("JPY", "392", 0),
	New("KES", "404", 2),
	New("KGS", "417", 2),
	New("KHR", "116", 2),
	New("KMF", "174", 0),
	New("KPW", "408", 2),
	New("KRW", "410", 0),
	New("KWD", "414", 3),
	New("KYD", "136", 2),
	New("KZT", "398", 2),
	New("LAK", "418", 2),
	New("LBP", "422", 2),
	New("LKR", "144", 2),
	New("LRD", "430", 2),
	New("LSL", "426", 2),
	New("LYD", "434", 3),
	New("MAD", "504", 2),
	New("MDL", "498", 2),
	New("MGA", "969", 2),
	New("MKD", "807", 2),
	New("MMK", "104", 2),
	New("MNT", "496", 2),
	New("MOP", "446", 2),
	New("MRU", "929", 2),
	New("MUR", "480", 2),
	New("MVR", "462", 2),
	New("MWK", "454", 2),
	New("MXN", "484", 2),
	New("MXV", "979", 2),
	New("MYR", "458", 2),
	New("MZN", "943", 2),
	New("NAD", "516", 2),
	New("NGN", "566", 2),
	New("NIO", "558", 2),
	New("NOK", "578", 2),
	New("NPR", "524", 2),
	New("NZD", "554", 2),
	New("OMR", "512", 3),
	New("PAB", "590", 2),
	New("PEN", "604", 2),
	New("PGK", "598", 2),
	New("PHP", "608", 2),
	New("PKR", "586", 2),
	New("PLN", "985", 2),
	New("PYG", "600", 0),
	New("QAR", "634", 2),
	New("RON", "946", 2),
	New("RSD", "941", 2),
	New("RUB", "643", 2),
	New("RWF", "646", 0),
	New("SAR", "682", 2),
	New("SBD", "090", 2),
	New("SCR", "690", 2),
	New("SDG", "938", 2),
	New("SEK", "752", 2),
	New("SGD", "702", 2),
	New("SHP", "654", 2),
	New("SLL", "694", 2),
	New("SOS", "706", 2),
	New("SRD", "968", 2),
	New("SSP", "728", 2),
	New("STN", "930", 2),
	New("SVC", "222", 2),
	New("SYP", "760", 2),
	New("SZL", "748", 2),
	New("THB", "764", 2),
	New("TJS", "972", 2),
	New("TMT", "934", 2),
	New("TND", "788", 3),
	New("TOP", "776", 2),
	New("TRY", "949", 2),
	New("TTD", "780", 2),
	New("TWD", "901", 2),
	New("TZS", "834", 2),
	New("UAH", "980", 2),
	New("UGX", "800", 0),
	New("USD", "840", 2),
	New("USN", "997", 2),
	New("UYI", "940", 0),
	New("UYU", "858", 2),
	New("UYW", "927", 4),
	New("UZS", "860", 2),
	New("VES", "928", 2),
	New("VND", "704", 0),
	New("VUV", "548", 0),
	New("WST", "882", 2),
	New("XAF", "950", 0),
	New("XCD", "951", 2),
	New("XOF", "952", 0),
	New("XPF", "953", 0),
	New("YER", "886", 2),
	New("ZAR", "710", 2),
	New("ZMW", "967", 2),
	New("ZWL", "932", 2),
}
